sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Common/AnalyzerConfiguration.cs

```diff
@@ -0,0 +1,96 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Xml.Linq;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Common
+{
+    public class AnalyzerConfiguration
+    {
+        private static readonly Lazy<HotspotConfiguration> hotspotConfiguration =
+            new Lazy<HotspotConfiguration>(() => new HotspotConfiguration());
+
+        private static readonly Lazy<IAnalyzerConfiguration> alwaysEnabledConfiguration =
+            new Lazy<IAnalyzerConfiguration>(() => new AlwaysEnabledConfiguration());
+
+        public static IAnalyzerConfiguration Hotspot =>
+            hotspotConfiguration.Value;
+
+        public static IAnalyzerConfiguration AlwaysEnabled =>
+            alwaysEnabledConfiguration.Value;
+
+        private class AlwaysEnabledConfiguration : IAnalyzerConfiguration
+        {
+            public void Initialize(AnalyzerOptions options) { }
+
+            public bool IsEnabled(string ruleKey) => true;
+        }
+
+        private class HotspotConfiguration : IAnalyzerConfiguration
+        {
+            private HashSet<string> enabledRules;
+
+            // This class is a singleton and there could be many rules that check if they are
+            // enabled simultaneously and since the XML loading is an IO operation (e.g. slow)
+            // we lock until it completes to prevent rules from wrongly deciding that they are
+            // disabled while the XML is loaded.
+            private static readonly object @lock = new object();
+
+            public bool IsEnabled(string ruleKey) =>
+                enabledRules.Contains(ruleKey);
+
+            public void Initialize(AnalyzerOptions options)
+            {
+                if (enabledRules != null)
+                {
+                    return;
+                }
+
+                lock (@lock)
+                {
+                    if (enabledRules != null)
+                    {
+                        return;
+                    }
+
+                    enabledRules = new HashSet<string>();
+
+                    var sonarLintAdditionalFile = options.AdditionalFiles
+                        .FirstOrDefault(f => ParameterLoader.ConfigurationFilePathMatchesExpected(f.Path));
+                    if (sonarLintAdditionalFile != null)
+                    {
+                        var sonarLintXml = XDocument.Load(sonarLintAdditionalFile.Path);
+
+                        var rulesInXml = sonarLintXml.Descendants("Rule")
```

Comment:
It's probably minor but I would add a comment here to explain that we expect the `SonarLint.xml` file to contain only rules that are active. At least it there is a change later on, when reading this it will be obvious.